grpc SendToRouteSync Error

### Background

When trying to send to a route via the grpc api function SendToRouteSync with nodejs this error is thrown. I most likely didn't construct the route correctly, however lnd crashes after I try to execute this.

When no routes are provided, a normal error message is thrown.

panic: runtime error: index out of range

`goroutine 5484 [running]:
github.com/lightningnetwork/lnd/routing.NewRouteFromHops(0x0, 0x39ab0c0300000000, 0x53389d373e524d77, 0xda0e8db3da12ea27, 0x5c1752df22119dc1, 0xf84a1ac44e, 0x1a90210, 0x0, 0x0, 0x0)
	/home/ubuntu/gocode/src/github.com/lightningnetwork/lnd/routing/pathfind.go:359 +0x256
main.(*rpcServer).unmarshallRoute(0xc0002e4180, 0xc010df9900, 0xc010d79aa8, 0xc010db7ee0, 0x3, 0x3)
	/home/ubuntu/gocode/src/github.com/lightningnetwork/lnd/rpcserver.go:3851 +0x2cd
main.(*rpcServer).SendToRouteSync(0xc0002e4180, 0x1281160, 0xc011157230, 0xc010df9800, 0xc0002e4180, 0xc010db7ac0, 0xc0001b3260)
	/home/ubuntu/gocode/src/github.com/lightningnetwork/lnd/rpcserver.go:2751 +0x111
github.com/lightningnetwork/lnd/lnrpc._Lightning_SendToRouteSync_Handler.func1(0x1281160, 0xc011157230, 0xf5a600, 0xc010df9800, 0x1, 0x1, 0x0, 0x0)
	/home/ubuntu/gocode/src/github.com/lightningnetwork/lnd/lnrpc/rpc.pb.go:6777 +0x89
github.com/lightningnetwork/lnd/macaroons.(*Service).UnaryServerInterceptor.func1(0x1281160, 0xc011157230, 0xf5a600, 0xc010df9800, 0xc010db7a40, 0xc010db7ac0, 0xeb1e80, 0x1a90210, 0xfdc940, 0xc011636400)
	/home/ubuntu/gocode/src/github.com/lightningnetwork/lnd/macaroons/service.go:125 +0x153
github.com/lightningnetwork/lnd/lnrpc._Lightning_SendToRouteSync_Handler(0xffa9a0, 0xc0002e4180, 0x1281160, 0xc011157230, 0xc011714c30, 0xc0002cbaa0, 0x0, 0x0, 0xc01140cc80, 0x48)
	/home/ubuntu/gocode/src/github.com/lightningnetwork/lnd/lnrpc/rpc.pb.go:6779 +0x158
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00044d380, 0x12886a0, 0xc011080600, 0xc011636400, 0xc0002d8ea0, 0x1a4d8b0, 0x0, 0x0, 0x0)
	/home/ubuntu/gocode/pkg/mod/google.golang.org/grpc@v1.16.0/server.go:935 +0x433
google.golang.org/grpc.(*Server).handleStream(0xc00044d380, 0x12886a0, 0xc011080600, 0xc011636400, 0x0)
	/home/ubuntu/gocode/pkg/mod/google.golang.org/grpc@v1.16.0/server.go:1139 +0xd58
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc010b5c9d0, 0xc00044d380, 0x12886a0, 0xc011080600, 0xc011636400)
	/home/ubuntu/gocode/pkg/mod/google.golang.org/grpc@v1.16.0/server.go:683 +0x9f
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/ubuntu/gocode/pkg/mod/google.golang.org/grpc@v1.16.0/server.go:681 +0xa1`

### Your environment

* lnd version 0.5.1-beta commit=v0.5.1-beta-138-g5451211d1947de5b2376aff5eb39c6e9f969cbbb
* Ubuntu 16.04 4.4.0-1072-aws
* neutrino

### Steps to reproduce

```
ln.addInvoice({value:

 1}, (err, invoice) => {
            if(err) throw err;

            console.log(invoice.r_hash.toString('hex'));
            
            ln.sendToRouteSync({payment_hash_string: invoice.r_hash.toString('hex'), routes: testroute.hops}, (err, response) => {
                if(err) throw err;
                console.log(response);
            });

        });
```

### Expected behaviour

Detailed error message for how the hops were wrong

--- a/routing/pathfind.go
+++ b/routing/pathfind.go
@@ -326,9 +326,12 @@ func newRoute(amtToSend, feeLimit lnwire.MilliSatoshi, sourceVertex Vertex,
 	}
 
 	// With the base routing data expressed as hops, build the full route
-	newRoute := NewRouteFromHops(
+	newRoute, err := NewRouteFromHops(
 		nextIncomingAmount, totalTimeLock, sourceVertex, hops,
 	)
+	if err != nil {
+		return nil, err
+	}
 
 	// Invalidate this route if its total fees exceed our fee limit.
 	if newRoute.TotalFees > feeLimit {
@@ -344,7 +347,11 @@ func newRoute(amtToSend, feeLimit lnwire.MilliSatoshi, sourceVertex Vertex,
 // information to perform the payment. It infers fee amounts and populates the
 // node, chan and prev/next hop maps.
 func NewRouteFromHops(amtToSend lnwire.MilliSatoshi, timeLock uint32,
-	sourceVertex Vertex, hops []*Hop) *Route {
+	sourceVertex Vertex, hops []*Hop) (*Route, error) {
+
+	if len(hops) == 0 {
+		return nil, ErrNoRouteHopsProvided
+	}
 
 	// First, we'll create a route struct and populate it with the fields
 	// for which the values are provided as arguments of this function.
@@ -369,7 +376,7 @@ func NewRouteFromHops(amtToSend lnwire.MilliSatoshi, timeLock uint32,
 		route.chanIndex[hop.ChannelID] = struct{}{}
 	}
 
-	return route
+	return route, nil
 }
 
 // Vertex is a simple alias for the serialization of a compressed Bitcoin

--- a/routing/pathfind_test.go
+++ b/routing/pathfind_test.go
@@ -1903,3 +1903,15 @@ func assertExpectedPath(t *testing.T, path []*channeldb.ChannelEdgePolicy,
 		}
 	}
 }
+
+// TestNewRouteFromEmptyHops tests that the NewRouteFromHops function returns an
+// error when the hop list is empty.
+func TestNewRouteFromEmptyHops(t *testing.T) {
+	t.Parallel()
+
+	var source Vertex
+	_, err := NewRouteFromHops(0, 0, source, []*Hop{})
+	if err != ErrNoRouteHopsProvided {
+		t.Fatalf("expected empty hops error: instead got: %v", err)
+	}
+}

--- a/routing/router_test.go
+++ b/routing/router_test.go
@@ -421,10 +421,13 @@ func TestChannelUpdateValidation(t *testing.T) {
 		},
 	}
 
-	route := NewRouteFromHops(
+	route, err := NewRouteFromHops(
 		lnwire.MilliSatoshi(10000), 100,
 		NewVertex(ctx.aliases["a"]), hops,
 	)
+	if err != nil {
+		t.Fatalf("unable to create route: %v", err)
+	}
 
 	// Set up a channel update message with an invalid signature to be
 	// returned to the sender.

--- a/rpcserver.go
+++ b/rpcserver.go
@@ -3848,12 +3848,15 @@ func (r *rpcServer) unmarshallRoute(rpcroute *lnrpc.Route,
 		prevNodePubKey = routeHop.PubKeyBytes
 	}
 
-	route := routing.NewRouteFromHops(
+	route, err := routing.NewRouteFromHops(
 		lnwire.MilliSatoshi(rpcroute.TotalAmtMsat),
 		rpcroute.TotalTimeLock,
 		sourceNode.PubKeyBytes,
 		hops,
 	)
+	if err != nil {
+		return nil, err
+	}
 
 	return route, nil
 }

